loky/backend/resource_tracker.py

```diff
@@ -0,0 +1,277 @@
+            raise ValueError('name too long')
+        nbytes = os.write(self._fd, msg)
+        assert nbytes == len(msg)
+
+
+_resource_tracker = ResourceTracker()
+ensure_running = _resource_tracker.ensure_running
+register = _resource_tracker.register
+unregister = _resource_tracker.unregister
+getfd = _resource_tracker.getfd
+
+
+def main(fd, verbose=0):
+    '''Run semaphore tracker.'''
+    # protect the process from ^C and "killall python" etc
+    signal.signal(signal.SIGINT, signal.SIG_IGN)
+    signal.signal(signal.SIGTERM, signal.SIG_IGN)
+
+    if _HAVE_SIGMASK:
+        signal.pthread_sigmask(signal.SIG_UNBLOCK, _IGNORED_SIGNALS)
+
+    for f in (sys.stdin, sys.stdout):
+        try:
+            f.close()
+        except Exception:
+            pass
+
+    if verbose:  # pragma: no cover
+        sys.stderr.write("Main semaphore tracker is running\n")
+        sys.stderr.flush()
+
+    cache = set()
+    try:
+        # keep track of registered/unregistered resources
+        with os.fdopen(fd, 'rb') as f:
+            for line in f:
+                try:
+                    cmd, name, rtype = line.strip().decode('ascii').split(':')
+                    cleanup_func = _CLEANUP_FUNCS.get(rtype, None)
+                    if cleanup_func is None:
+                        raise ValueError('Cannot register for automatic '
+                                         'cleanup: unknown resource type {}'
+                                         .format(name, rtype))
+
+                    if cmd == 'REGISTER':
+                        cache.add((name, cleanup_func))
+                        if verbose:  # pragma: no cover
+                            sys.stderr.write("[ResourceTracker] register {}"
+                                             " {}\n" .format(rtype, name))
+                            sys.stderr.flush()
+                    elif cmd == 'UNREGISTER':
+                        cache.remove((name, cleanup_func))
+                        if verbose:  # pragma: no cover
+                            sys.stderr.write("[ResourceTracker] unregister {}"
+                                             " {}: cache({})\n"
+                                             .format(name, rtype, len(cache)))
+                            sys.stderr.flush()
+                    elif cmd == 'PROBE':
+                        pass
+                    else:
+                        raise RuntimeError('unrecognized command %r' % cmd)
+                except BaseException:
+                    try:
+                        sys.excepthook(*sys.exc_info())
+                    except BaseException:
+                        pass
+    finally:
+        # all processes have terminated; cleanup any remaining resources
+        if cache:
+            try:
+                warnings.warn('resource_tracker: There appear to be %d '
+                              'leaked resources to clean up at shutdown' %
+                              len(cache))
```

Comment:
I think that it would be better to let the user know what type of resource he is leaking, so he can fix the leak :)

For this, I would use a dict with a set per `rtype` for the `cache`:
```
cache = {rtype: set() for rtype in _CLEANUP_FUNCS.keys()}
```

And then you `register/unregister` only names in `cache[rtype]` and at the end, you just loop over the different `rtype`/`cleanup` and their respective `cache`.

Reply:
Yes, this is definitely something to consider. I was waiting for your input on that topic.